### PR TITLE
Add ASP.NET sample applications

### PR DESCRIPTION
**Pull request by jpassing**
Add two ASP.NET sample applications that can be used to test
different deployment methods:

* net4/ contains a ASP.NET MVC 5 app that can be run in IIS/Windows
* netcore/ contains an ASP.NET MVC Core app that can be run on Linux

Both apps are equivalent in terms of functionality and structure

Planning to update the following tutorials to use these new sample applications:
 * https://cloud.google.com/solutions/creating-cicd-pipeline-vsts-kubernetes-engine
 * https://cloud.google.com/solutions/creating-cicd-pipeline-vsts-compute-engine
 * https://cloud.google.com/solutions/creating-a-cicd-pipeline-azure-pipelines-cloud-run

### CODE REVIEW

File: applications/clouddemo/README.md
Comment (amanda-tarafa) [inline applications/clouddemo/README.md:6]: and also on Windows. Core runs everywhere, this might give the impression that Core only runs on Linux.

File: applications/clouddemo/README.md
Comment (amanda-tarafa) [inline applications/clouddemo/README.md:3]:
```suggestion
This folder contains sample ASP.NET applications:
```

File: applications/clouddemo/net4/CloudDemo.Mvc/Controllers/HomeController.cs
Comment (amanda-tarafa) [inline applications/clouddemo/net4/CloudDemo.Mvc/Controllers/HomeController.cs:42]: nit: Remove the empty comment lines, here and everywhere.

**Review by amanda-tarafa — CHANGES_REQUESTED** [review]
I've left a few comments but I have a couple of general concerns:

- These don't have tests, at the very least we should check that they build. Ideally, we should have something like the [HelloWord AppEngine sample]( https://github.com/GoogleCloudPlatform/dotnet-docs-samples/tree/master/appengine/flexible/HelloWorld) (see the deployTests.ps1 and the runTests.ps1 scripts)

- It seems to me that both of these projects contain all the files that the VisualStudio template generate. Most of those are not relevant to what wants to be shown and it might end up confusing users. We tend to provide minimum projects, removing everything superflous. This is not a hard requirement though.

- All those code and config files that are generated by VisualStudio and the are not modified by the sample have the license, including the Google Copyright. Not sure if we can claim copyright of files generated by VisualStudio that we haven't modified. Might be worth to check with someone at legal.

- Can you paste a print screen of what the HomePage looks like? Is the css the one that VisualStudio generates? If that's the case, could we either remove the css or have one of our own?

**Comment by jpassing** [comment]
> * These don't have tests, at the very least we should check that they build. Ideally, we should have something like the [HelloWord AppEngine sample](https://github.com/GoogleCloudPlatform/dotnet-docs-samples/tree/master/appengine/flexible/HelloWorld) (see the deployTests.ps1 and the runTests.ps1 scripts)

I do not think there's anything to test, but I could have kokoro at least perform a build. Do you have an example for how that would work with Windows?

> * It seems to me that both of these projects contain all the files that the VisualStudio template generate. Most of those are not relevant to what wants to be shown and it might end up confusing users. We tend to provide minimum projects, removing everything superflous. This is not a hard requirement though.

I'd like to keep the projects close to what the general conventions are. I could get rid of some files, but I am afraid that it'll just make it more difficult for users.

> * All those code and config files that are generated by VisualStudio and the are not modified by the sample have the license, including the Google Copyright. Not sure if we can claim copyright of files generated by VisualStudio that we haven't modified. Might be worth to check with someone at legal.

Not sure what you'd like me to do here. There's always going to be generated source code when using .NET, and our OSS policies require me to add an APL license header everywhere.

> * Can you paste a print screen of what the HomePage looks like? Is the css the one that VisualStudio generates? If that's the case, could we either remove the css or have one of our own?

Here's what the app looks like, the CSS is based on MDL:

Capture

**Comment by amanda-tarafa** [comment]
> > * These don't have tests, at the very least we should check that they build. Ideally, we should have something like the [HelloWord AppEngine sample](https://github.com/GoogleCloudPlatform/dotnet-docs-samples/tree/master/appengine/flexible/HelloWorld) (see the deployTests.ps1 and the runTests.ps1 scripts)
> 
> I do not think there's anything to test, but I could have kokoro at least perform a build. Do you have an example for how that would work with Windows?
> 

You can test that it actually deploys and that when you ping the host you get back the expected page, or at least a 200. The link above does something like that, although their tests are more complex than what I'm describing.
For testing that it builds only, add a runTests.ps1 script like [this one](https://github.com/GoogleCloudPlatform/dotnet-docs-samples/blob/master/functions/slack/runTests.ps1) for Core and add the following two lines at the start for Windows:

```
Import-Module ..\..\BuildTools.psm1 -DisableNameChecking

Require-Platform Win*
```
This repo CI will pick that up and run the tests for you. Do try to check that the app deploys and responds correctly though. It's very easy for these to become stale.

> > * It seems to me that both of these projects contain all the files that the VisualStudio template generate. Most of those are not relevant to what wants to be shown and it might end up confusing users. We tend to provide minimum projects, removing everything superflous. This is not a hard requirement though.
> 
> I'd like to keep the projects close to what the general conventions are. I could get rid of some files, but I am afraid that it'll just make it more difficult for users.
> 

So, VS templates are not conventions per se, they are the current recommendations by MS, but those change all the time when features are added. What happens when we add a project based in templates as these are, is that they become stale very quickly, and users that don't know better end up using old recommendations. Also, the most inexperienced users will have a hard time distinguishing what is the code relevant to the sample and what is boilerplate code. That's why we strip everything we can, so the sample is really focused on what we want to show. See [this one](https://github.com/googleapis/google-api-dotnet-client/tree/master/Src/Support/Google.Apis.Auth.AspNetCore3.IntegrationTests) for instace (not even a css). As I said earlier this is not a hard requirement, but it will definetely help both users and future maintenance of the sample code.

> > * All those code and config files that are generated by VisualStudio and the are not modified by the sample have the license, including the Google Copyright. Not sure if we can claim copyright of files generated by VisualStudio that we haven't modified. Might be worth to check with someone at legal.
> 
> Not sure what you'd like me to do here. There's always going to be generated source code when using .NET, and our OSS policies require me to add an APL license header everywhere.
> 

No, there's not always going to be generated code when using .NET.
In any case, my suggestion is that you check with legal, I have no idea if my concern makes sense or not.

**Comment by jpassing** [comment]
> > > * All those code and config files that are generated by VisualStudio and the are not modified by the sample have the license, including the Google Copyright. Not sure if we can claim copyright of files generated by VisualStudio that we haven't modified. Might be worth to check with someone at legal.
> > 
> > 
> > Not sure what you'd like me to do here. There's always going to be generated source code when using .NET, and our OSS policies require me to add an APL license header everywhere.
> 
> No, there's not always going to be generated code when using .NET.
> In any case, my suggestion is that you check with legal, I have no idea if my concern makes sense or not.

This repository contains plenty of projects with VS-generated files (Startup.cs, Program.cs, *.sln, *.csproj, etc), so if this is really a concern, the entire repository ought to be checked for this.

**Comment by amanda-tarafa** [comment]
> > > > * All those code and config files that are generated by VisualStudio and the are not modified by the sample have the license, including the Google Copyright. Not sure if we can claim copyright of files generated by VisualStudio that we haven't modified. Might be worth to check with someone at legal.
> > > 
> > > 
> > > Not sure what you'd like me to do here. There's always going to be generated source code when using .NET, and our OSS policies require me to add an APL license header everywhere.
> > 
> > 
> > No, there's not always going to be generated code when using .NET.
> > In any case, my suggestion is that you check with legal, I have no idea if my concern makes sense or not.
> 
> This repository contains plenty of projects with VS-generated files (Startup.cs, Program.cs, *.sln, *.csproj, etc), so if this is really a concern, the entire repository ought to be checked for this.

Yes, but lot of those files have changes in them that are specific to the sample (*.sln and *.csproj all of them for sure as do yours, and most Startups.cs and Programs.cs that I've seen as well). In any case and as I said before, I'm not sure my concern is a valid one, that's why I only expressed my comment around this as a suggestion.

**Comment by jpassing** [comment]
I removed/combined a few non-essential files.

**Comment by amanda-tarafa** [comment]
Thanks! Let me know when you have added the runTests.ps1 file with at least the build bits. Also, let me know if I can help with that.

**Comment by jpassing** [comment]
I added `runTests.ps1` files for both projects. 

Unfortunately, the build is currently failing for other reasons.

**Comment by amanda-tarafa** [comment]
One of your runs is [failing](https://source.cloud.google.com/results/invocations/0e212a1e-2235-4a10-86de-3f99e72046ac/targets/github%2Fdotnet-docs-samples%2Fapplications%2Fclouddemo%2Fnet4/tests;group=runTests;test=runTests;row=1).
I'm leaving comments on the runTests.ps1 files now.

(And yes, there's somo flakinness on this repo, once your tests are green I can force merge, no worries).

File: applications/clouddemo/netcore/runTests.ps1
Comment (amanda-tarafa) [inline applications/clouddemo/netcore/runTests.ps1:20]:
```suggestion
dotnet build 2>&1 | %{ "$_" }
```
This should be enough. Also, we need redirecting stderr to stdout so that we get all the error messages properly. That's what  the `2>&1 | %{ "$_" }`.

File: applications/clouddemo/net4/runTests.ps1
Comment (amanda-tarafa) [inline applications/clouddemo/net4/runTests.ps1:31]:
```suggestion
dotnet build 2>&1 | %{ "$_" }
```
Same here, don't think there's a need for all the rest.

**Review by amanda-tarafa — CHANGES_REQUESTED** [review]

**Comment by amanda-tarafa** [comment]
Still [failing](https://source.cloud.google.com/results/invocations/159169c1-03be-408b-ae30-9648ebe91563/targets/github%2Fdotnet-docs-samples%2Fapplications%2Fclouddemo%2Fnet4/tests).

Any particular reason you are not building the apps in the same way that almost everything else is built on this repo, and as per I suggested in my last two comments? Just curious, `dotnet build` (plus the stdout redirection bits) should be enough for everything. We don't need `dotnet publish` and also we don't need to use `msbuild`. In general, I would favor homogeneity of the repo over anything else, unless there's a compelling reason.

**Comment by jpassing** [comment]
What's wrong with using `dotnet publish` instead of `build`? At least that ensures that the publishing profile is used well.

The project in the `net4` folder is a .NET Framework project with a non-SDK style csproj file, so it must use `msbuild`.

**Comment by amanda-tarafa** [comment]
> What's wrong with using `dotnet publish` instead of `build`? At least that ensures that the publishing profile is used as well.

Good point about `publish` checking the publishing profile. Nothing wrong with using `publish` as long as there's a reason for it (as there is), else, we try to keep things homogenous, this is a big repo with lots of contributors.

> The project in the `net4` folder is a .NET Framework project with a non-SDK style csproj file, so it must use `msbuild`.

Ah yes, true.

It's still failing though. I'll rerun Kokoro now, to make sure that we are not looking at old failures.

**Comment by amanda-tarafa** [comment]
Still [failing](https://source.cloud.google.com/results/invocations/05f20965-4a6b-416c-9946-a51229f87392/targets/github%2Fdotnet-docs-samples%2Fapplications%2Fclouddemo%2Fnet4/tests).

```
MSBUILD : error MSB1008: Only one project can be specified.
Switch: CloudDemo.Mvc\CloudDemo.Mvc.csproj
```

With a quick search, it seems that this error mostly happens if there are spaces in the path, but there are none in this instance, so not sure what this might be. Let me know if I can help.

**Comment by jpassing** [comment]
Fixed the `msbuild` command so that it's now compatible with the older msbuild version that Kokoro uses.

I also force-added the publishing profiles -- they are needed for the build, but the default `.gitignore` in this repo excluded them.

**Review by amanda-tarafa — APPROVED** [review]
LGTM!
I'll have to force merge, we require 2 approvals, one from the sample owner (that'd be you in this case) and one from a repo owner (me). Let me know when you are ready.

**Comment by jpassing** [comment]
Thanks, Amanda. Let's merge then.

**Pull request merged**
